client-lite/src/util/do_curl_multi_operation.cpp

```diff
@@ -0,0 +1,247 @@
+
+        _RemoveHandle(easyHandle, _handlesToAdd);
+
+        auto pExistingWrappedHandle = _activeHandles.Get(easyHandle);
+        if (pExistingWrappedHandle == nullptr)
+        {
+            return;
+        }
+        wrappedHandleToBeDeleted = *pExistingWrappedHandle;
+
+        if (std::find(std::begin(_handlesToRemove), std::end(_handlesToRemove), easyHandle) == _handlesToRemove.end())
+        {
+            _handlesToRemove.emplace_back(easyHandle);
+        }
+        _cv.notify_one();
+    }
+
+    if (wrappedHandleToBeDeleted)
+    {
+        // outside the lock, wait for the handle to be removed from _activeHandles
+        (void)wrappedHandleToBeDeleted->inactiveSignal.Wait(std::chrono::minutes(30));
+    }
+}
+
+void CurlMultiOperation::_WorkerThread()
+{
+    while (true)
+    {
+        std::unique_lock<std::mutex> lock{_mutex};
+        if (!_fKeepRunning)
+        {
+            break;
+        }
+
+        for (const auto& h : _handlesToAdd)
+        {
+            _activeHandles.Add(h, _multiHandle);
+        }
+        _handlesToAdd.clear();
+
+        for (const auto& h : _handlesToRemove)
+        {
+            _activeHandles.Remove(h, _multiHandle);
+        }
+        _handlesToRemove.clear();
+
+        if (_activeHandles.Empty())
+        {
+            _cv.wait(lock, [this](){ return !_fKeepRunning || !_handlesToAdd.empty(); });
+        }
+        else
+        {
+            lock.unlock();
+            _PeformOperations();
+        }
+    }
+}
+
+// Returns when there are no more running handles OR after running for approximately 2s
+void CurlMultiOperation::_PeformOperations()
+{
+    int numRunningHandles = 0;
+    const auto startTime = std::chrono::steady_clock::now();
+    std::chrono::steady_clock::duration elapsed {};
+    do
+    {
+        long curlTimeoutMsecs;
+        curl_multi_timeout(_multiHandle, &curlTimeoutMsecs);
+        if (curlTimeoutMsecs < 0)
+        {
+            curlTimeoutMsecs = 1000;
+        }
+
+        fd_set fdsRead;
+        fd_set fdsWrite;
+        fd_set fdsException;
+        int maxfd = -1;
+        FD_ZERO(&fdsRead);
+        FD_ZERO(&fdsWrite);
+        FD_ZERO(&fdsException);
+
+        int selectResult;
+        CURLMcode mc = curl_multi_fdset(_multiHandle, &fdsRead, &fdsWrite, &fdsException, &maxfd);
+        if (maxfd == -1)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(100)); // no relevant handles to poll, short sleep
+            selectResult = 0;
+        }
+        else
+        {
+            struct timeval timeout;
+            timeout.tv_sec = curlTimeoutMsecs / 1000;
+            timeout.tv_usec = (curlTimeoutMsecs % 1000) * 1000;
+            selectResult = select(maxfd + 1, &fdsRead, &fdsWrite, &fdsException, &timeout);
+        }
+
+        switch (selectResult)
+        {
+            case -1:
+                DO_ASSERT(false); // TODO(shishirb) how to handle?
+                break;
+
+            case 0:
```

Comment:
I think you might have mentioned this when we went over this, but for case 0: it indicates the process timed out? Should a comment be added on why we leave this case empty?

Reply:
0 = timeout, yes. Note this case isn't empty but falls through to call curl_multi_perform. I've added some comments in the next iteration.